suppressPackageStartupMessages(library(optparse))
suppressPackageStartupMessages(library(futile.logger))
suppressPackageStartupMessages(library(pROC))
suppressPackageStartupMessages(library(PureCN))
### Parsing command line ------------------------------------------------------

option_list <- list(
    make_option(c("--rds"), action="store", type="character", default=NULL,
        help="Input: PureCN output RDS file"),
    make_option(c("--vcf"), action = "store", type = "character", default = NULL,
                help = "Input: VCF file containing SOMATIC calls"),
    make_option(c("--callable"), action="store", type="character", default=NULL,
        help="TMB: File parsable by rtracklayer specifying all callable regions, e.g. generated by GATK CallableLoci"),
    make_option(c("--keepflagged"), action = "store_true", default=FALSE,
        help = "Keep variants flagged by predictSomatic [default %default]"),
    make_option(c("--keepdb"), action = "store_true", default=FALSE,
        help = "Keep variants in dbSNP [default %default]"),
    make_option(c("--callcutoff"), action="store", type="double", default=0.8,
        help="Posterior probability cutoff to include variants [default %default]"),
    make_option(c("--sgzdir"), action="store", type="character", default="",
        help="Directory with SGZ files [default %default]."),
    make_option(c("--out"), action="store", type="character", default=NULL,
        help="File name prefix to which results should be written"),
    make_option(c("-v", "--version"), action="store_true", default=FALSE,
        help="Print PureCN version"),
    make_option(c("-f", "--force"), action="store_true", default=FALSE,
        help="Overwrite existing files")
)

opt <- parse_args(OptionParser(option_list=option_list))

if (opt$version) {
    message(as.character(packageVersion("PureCN")))
    q(status=1)
}

# Parse input rds
infileRds <- opt$rds
if (is.null(infileRds)) stop("Need --rds")
infileRds <- normalizePath(infileRds, mustWork=TRUE)

# sgz directory
sgzdir<-opt$sgzdir

# Parse both BED files restricting covered region
callableFile <- opt$callable
callable <- NULL
if (!is.null(callableFile)) {
    suppressPackageStartupMessages(library(rtracklayer))
    callableFile <- normalizePath(callableFile, mustWork=TRUE)
    flog.info("Reading %s...", callableFile)
    callable <- GRanges(import(callableFile))
}

### Run PureCN ----------------------------------------------------------------

res <- readCurationFile(infileRds)
sampleid <- res$input$sampleid

.getOutPrefix <- function(opt, infile, sampleid) {
    out <- opt[["out"]]
    if (is.null(out)) {
        if (!is.null(infile) && file.exists(infile)) {
            outdir <- dirname(infile)
            prefix <- sampleid
        } else {
            stop("Need --out")
        }
    } else {
        outdir <- dirname(out)
        prefix <- basename(out)
    }
    outdir <- normalizePath(outdir, mustWork=TRUE)
    outPrefix <- file.path(outdir, prefix)
}
outPrefix <- .getOutPrefix(opt, infileRds, sampleid)

flog.info("Reading %s...", basename(opt$vcf))
vcf <- PureCN:::.readAndCheckVcf(opt$vcf, genome = genome(res$input$vcf)[1])
loh <- callLOH(res)
sgz_cna <- data.frame(
    CHR = loh$chr,
    segStart = floor(loh$start),
    segEnd = ceiling(loh$end),
    mafPred = loh$maf.expected,
    CN = loh$C,
    segLR = loh$seg.mean,
    segMAF = loh$maf.observed,
    numMAtumorPred = loh$M,
    numLRProbes = loh$num.mark,
    numAFProbes = loh$num.snps,
    purity = res$results[[1]]$purity,
    baseLevel = res$results[[1]]$total.ploidy
)
sgz_cna <- sgz_cna[which(sgz_cna$numAFProbes>=1),]

flog.info("Classifying germline vs. somatic...")

df <- data.frame(matrix(ncol = 8, nrow = 0))
colnames(df)<-c("#sample","mutation","frequency","depth","pos","status","strand","effect")

chr<-as.vector(seqnames(vcf))
pos<-as.vector(start(vcf))
AF<-as.vector(info(vcf)$AF)
DP<-as.vector(info(vcf)$DP)
ANN<-as.vector(info(vcf)$ANN)

for (i in 1:length(vcf)){
  if (startsWith(unlist(strsplit(ANN[[i]],"[|]"))[2], "missense")){
     mutation<-paste(unlist(strsplit(ANN[[i]],"[|]"))[4],':',unlist(strsplit(ANN[[i]],"[|]"))[7],':',paste(unlist(strsplit(ANN[[i]],"[|]"))[10:11],collapse="_"),sep="")
     frequency<-AF[[i]]
     depth<-DP[[i]]
     position<-paste(chr[[i]],pos[[i]],sep=":")
     df[i,1:8]<-c(sampleid,mutation,frequency,depth,position,'unknown','*','missense')
  }
  if (startsWith(unlist(strsplit(ANN[[i]],"[|]"))[2], "frameshift")){
    mutation<-paste(unlist(strsplit(ANN[[i]],"[|]"))[4],':',unlist(strsplit(ANN[[i]],"[|]"))[7],':',paste(unlist(strsplit(ANN[[i]],"[|]"))[10:11],collapse="_"),sep="")
    frequency<-AF[[i]]
    depth<-DP[[i]]
    position<-paste(chr[[i]],pos[[i]],sep=":")
    df[i,1:8]<-c(sampleid,mutation,frequency,depth,position,'unknown','*','frameshift')
  }
  if (startsWith(unlist(strsplit(ANN[[i]],"[|]"))[2], "nonsense")){
    mutation<-paste(unlist(strsplit(ANN[[i]],"[|]"))[4],':',unlist(strsplit(ANN[[i]],"[|]"))[7],':',paste(unlist(strsplit(ANN[[i]],"[|]"))[10:11],collapse="_"),sep="")
    frequency<-AF[[i]]
    depth<-DP[[i]]
    position<-paste(chr[[i]],pos[[i]],sep=":")
    df[i,1:8]<-c(sampleid,mutation,frequency,depth,position,'unknown','*','nonsense')
  }
  if (startsWith(unlist(strsplit(ANN[[i]],"[|]"))[2], "synonymous")){
    mutation<-paste(unlist(strsplit(ANN[[i]],"[|]"))[4],':',unlist(strsplit(ANN[[i]],"[|]"))[7],':',paste(unlist(strsplit(ANN[[i]],"[|]"))[10:11],collapse="_"),sep="")
    frequency<-AF[[i]]
    depth<-DP[[i]]
    position<-paste(chr[[i]],pos[[i]],sep=":")
    df[i,1:8]<-c(sampleid,mutation,frequency,depth,position,'unknown','*','synonymous')
  }
}

sgz_mut<-df[complete.cases(df), ]

outfileMut <- paste0(outPrefix, "_sgz.mut_aggr.full.txt")
outfileCna <- paste0(outPrefix, "_sgz.cna.calls.txt")
write.table(sgz_cna, file = outfileCna, sep = "\t", row.names = FALSE,
    quote = FALSE)
write.table(sgz_mut, file = outfileMut, sep = "\t", row.names = FALSE,
    quote = FALSE)

system2(file.path(sgzdir, "fmiSGZ.py"), paste(outfileMut, outfileCna, "-o", outPrefix))
